agent/xds/listeners.go

```diff
@@ -512,7 +548,87 @@ func (s *Server) makeUpstreamListenerIgnoreDiscoveryChain(
 	return l, nil
 }
 
-func (s *Server) makeGatewayListener(name, addr string, port int, cfgSnap *proxycfg.ConfigSnapshot) (*envoy.Listener, error) {
+func (s *Server) makeTerminatingGatewayListener(name, addr string, port int, cfgSnap *proxycfg.ConfigSnapshot, token string) (*envoy.Listener, error) {
+	l := makeListener(name, addr, port)
+
+	tlsInspector, err := makeTLSInspectorListenerFilter()
+	if err != nil {
+		return nil, err
+	}
+	l.ListenerFilters = []envoylistener.ListenerFilter{tlsInspector}
+
+	// Make a FilterChain for each linked service
+	// Match on the cluster name,
+	for svc, _ := range cfgSnap.TerminatingGateway.ServiceGroups {
+		clusterName := connect.ServiceSNI(svc.ID, "", svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+		resolver, hasResolver := cfgSnap.TerminatingGateway.ServiceResolvers[svc]
+
+		clusterChain, err := s.sniFilterChainTerminatingGateway(name, clusterName, svc, cfgSnap)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make filter chain for cluster %q: %v", clusterName, err)
+		}
+		l.FilterChains = append(l.FilterChains, clusterChain)
+
+		// if there is a service-resolver for this service then also setup subset filter chains for it
+		if hasResolver {
+			// generate 1 filter chain for each service subset
+			for subsetName := range resolver.Subsets {
+				clusterName := connect.ServiceSNI(svc.ID, subsetName, svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+
+				clusterChain, err := s.sniFilterChainTerminatingGateway(name, clusterName, svc, cfgSnap)
+				if err != nil {
+					return nil, fmt.Errorf("failed to make filter chain for cluster %q: %v", clusterName, err)
+				}
+				l.FilterChains = append(l.FilterChains, clusterChain)
+			}
+		}
+	}
+
+	err = injectConnectFilters(cfgSnap, token, l, false)
```

Comment:
you're missing an `if err != nil { return nil, err }` here

Reply:
Fixed